rclcpp/src/rclcpp/node_interfaces/node_base.cpp

```diff
@@ -31,6 +32,75 @@ using rclcpp::exceptions::throw_from_rcl_error;
 
 using rclcpp::node_interfaces::NodeBase;
 
+namespace
+{
+/// This class bundles together the lifetime of the rcl_node_t handle with the
+/// lifetime of the RCL context. This ensures that the context will remain alive
+/// for as long as the rcl_node_t handle is alive.
+class NodeHandleWithContext
+{
+public:
+  /// The make function returns a std::shared_ptr<rcl_node_t> which is actually
+  /// an alias for a std::shared_ptr<NodeHandleWithContext>. There is no use for
+  /// accessing a NodeHandleWithContext instance directly, because its only job
+  /// is to couple together the lifetime of a context with the lifetime of a
+  /// node handle that depends on it.
+  static std::shared_ptr<rcl_node_t>
+  make(
+    rclcpp::Context::SharedPtr context,
+    std::shared_ptr<std::recursive_mutex> logging_mutex,
+    rcl_node_t * node_handle)
+  {
+    auto aliased_ptr = std::shared_ptr<NodeHandleWithContext>(
+      new NodeHandleWithContext(
+        std::move(context),
+        std::move(logging_mutex),
+        node_handle));
```

Comment:
Can you use `std::make_shared` here? I think that would be better, unless the aliased pointer requires two allocations as is done here.

Reply:
We can't use `std::make_shared` here because `NodeHandleWithContext` has a private constructor, and that unfortunately prohibits `std::make_shared` from being able to call it. If you try to side-step that by making `std::make_shared` a friend function of the class, then suddenly anyone can construct a `NodeHandleWithContext` anywhere simply by calling `std::make_shared<NodeHandleWithContext>`.

The aliasing pointer doesn't require the two allocations (I assume you're referring to [1] the allocation of the reference-counting block and [2] the allocation of the `NodeHandleWithContext` instance), it's just a matter of whether we want the code to self-document the way this class is meant to be used. Personally I think ensuring correctness is worth the very small additional allocation, but I suppose it's ultimately a subjective value judgment.

Reply:
As a side note, one method I sometimes use to side-step this limitation is to use move semantics, like this:

```
ClassWithPrivateConstructor my_instance;
return std::make_shared<ClassWithPrivateConstructor>(std::move(my_instance));
```

That can work if your class is movable, but we're explicitly disabling movability here because we have a destructor that does some special RAII work, and so we'd need to make a custom move constructor or else we could end up double-deleting.

Reply:
> We can't use std::make_shared here because NodeHandleWithContext has a private constructor, and that unfortunately prohibits std::make_shared from being able to call it

`NodeHandleWithContext` is declared and defined in a .cpp file and there's no reference to it in any header AFAIS, it doesn't seem to be a problem to make the constructor public IMO.
We can also mark the constructor with the `RCLCPP_LOCAL` attribute, so the symbol is hidden.

Reply:
When possible I like to put constraints even inside implementations, as reminders to the old maintainers and clues to the new maintainers about how an implementation detail is meant to work. I don't mind making a custom move constructor for this class if it's considered important to have a single allocation.

Reply:
With c8716ac we are now doing a single allocation for the `std::shared_ptr`.